### PR TITLE
feat(username): adds github username of the contributor

### LINKED ISSUE #39: Return `GitHub Name (@github-handle)` in list of contributors
This is a nice to have that would match the Ember Times "style guide"

### PR DESCRIPTION
**Pull request by abhilashlr**
closes #39

### CODE REVIEW

**Comment by ijlee2** [comment]
I wonder if the solution to getting the contributor names is indeed to make an additional request.

In [PR #37](https://github.com/ember-learn/whats-new-in-emberland/pull/37/files#r463693958), I noticed that we don't see the repo name or URL. I wondered if GitHub changed the payload structure over time. If so, we could update our code to get data from the new structure.

Before we look at that, though, I wanted to refactor/clean up the `index` route so that it's easier to see what's going on there. Perhaps we may be able to get the contributor names somehow by making a change in the route file.

In short, maybe we can wait on reviewing this PR until we merge #41 first, then look into cleaning up the `index` route. Would this be okay with you?

**Comment by abhilashlr** [comment]
> I wonder if the solution to getting the contributor names is indeed to make an additional request.
> 
> In [PR #37](https://github.com/ember-learn/whats-new-in-emberland/pull/37/files#r463693958), I noticed that we don't see the repo name or URL. I wondered if GitHub changed the payload structure over time. If so, we could update our code to get data from the new structure.
> 
> Before we look at that, though, I wanted to refactor/clean up the `index` route so that it's easier to see what's going on there. Perhaps we may be able to get the contributor names somehow by making a change in the route file.
> 
> In short, maybe we can wait on reviewing this PR until we merge #41 first, then look into cleaning up the `index` route. Would this be okay with you?

Yep, that works @ijlee2.

**Comment by abhilashlr** [comment]
@ijlee2 Rebased this with the master branch.

**Comment by ijlee2** [comment]
Thanks! Can you try rebasing once more? PRs will now need to pass lint and test checks.

**Comment by abhilashlr** [comment]
> Thanks! Can you try rebasing once more? PRs will now need to pass lint and test checks.

Oops! Done @ijlee2 👍

File: app/controllers/index.js [hunk 1]
```diff
@@ -54,11 +56,28 @@ export default class IndexController extends Controller {
...
+    let userPromises = [];
+    users.forEach((user) => {
+      userPromises.push(fetch(user.url).then((response) => response.json()));
+    });
+
+    let userDataResponses = await all(userPromises);
+
+    return users.map((user, index) => ({
+      ...user,
+      name: userDataResponses[index].name || "",
```
Comment (ijlee2) [inline app/controllers/index.js:69]: **suggestion (non-blocking):**

I think we can avoid setting a default value here (and let the consuming site default it however it needs to). I wonder if `name` is always defined.

```suggestion
      name: userDataResponses[index].name,
```

File: app/controllers/index.js [hunk 2]
```diff
@@ -54,11 +56,28 @@ export default class IndexController extends Controller {
...
   updateContributorsList(users) {
     const contributorsList = users
       .map(user => {
-        const { handle, profileLink } = user;
-        return `<a href="${profileLink}" rel="noopener noreferrer" target="_blank">${handle}</a>`;
+        const { handle, profileLink, name } = user;
+
+        const constructedName = name ? `${name} ` : '';
+
+        return `<a href="${profileLink}" rel="noopener noreferrer" target="_blank">${constructedName}(${handle})</a>`;
```
Comment (ijlee2) [inline app/controllers/index.js:80]: **suggestion (soft-blocking):**

When `name` doesn't exist, I think showing the handle without parentheses may look better.

Can we use an if statement to build display name?

```suggestion
        let displayName;

        if (name) {
          displayName = `${name} (${handle})`;
        } else {
          displayName = handle;
        }

        return `<a href="${profileLink}" rel="noopener noreferrer" target="_blank">${displayName}</a>`;
```

File: tests/unit/controllers/index-test.js [hunk 3]
```diff
@@ -140,18 +151,19 @@ module('Unit | Controller | index', function(hooks) {
...
     ]);
 
     assert.strictEqual(
       controller.contributorsList,
       [
-        '<a href="https://github.com/emberjs" rel="noopener noreferrer" target="_blank">@emberjs</a>',
-        '<a href="https://github.com/tomster" rel="noopener noreferrer" target="_blank">@tomster</a>',
-        '<a href="https://github.com/zoey" rel="noopener noreferrer" target="_blank">@zoey</a>'
+        '<a href="https://github.com/emberjs" rel="noopener noreferrer" target="_blank">(@emberjs)</a>',
+        '<a href="https://github.com/tomster" rel="noopener noreferrer" target="_blank">(@tomster)</a>',
```
Comment (ijlee2) [inline tests/unit/controllers/index-test.js:163]: If we decide to not use parentheses when `name` isn't present, these lines will need to be updated.

File: tests/unit/controllers/index-test.js [hunk 4]
```diff
@@ -140,18 +151,19 @@ module('Unit | Controller | index', function(hooks) {
...
+        '<a href="https://github.com/zoey" rel="noopener noreferrer" target="_blank">Zoey (@zoey)</a>'
```
Comment (ijlee2) [inline tests/unit/controllers/index-test.js:164]: **praise:**

Thanks for covering both test cases!

File: app/controllers/index.js [hunk 5]
```diff
@@ -17,22 +17,24 @@ export default class IndexController extends Controller {
     let users = await all(fetchRequests);
     users = this.identifyUsers(users);
     users = this.sortUsers(users);
+    users = await this.fetchAdditionalUserDetails(users);
```
Comment (ijlee2) [inline app/controllers/index.js:20]: **thought:**

I think we could switch the order of `sortUsers` and `fetchAdditionalUserDetails`, so that we have the option to sort in a different way once we know additional information about users.

I think, for now, it's perfectly fine to leave this line as is.

File: app/controllers/index.js [hunk 6]
```diff
@@ -17,22 +17,24 @@ export default class IndexController extends Controller {
...
 
       const isUser = (type === 'User');
       const isNotDuplicate = !accumulator.has(username);
 
       if (isUser && isNotDuplicate) {
         accumulator.set(username, {
           handle: `@${username}`,
-          profileLink: htmlUrl
+          profileLink: htmlUrl,
+          url
```
Comment (ijlee2) [inline app/controllers/index.js:37]: **suggestion (non-blocking):**

Without looking at the tests, I think it's a bit hard to understand the purpose of `url`. Maybe we rename it to `apiLink` to indicate that it's meant to be used in a request somehow?

If so, line 62 will need to be updated as well.

```suggestion
          apiLink: url
```

File: app/controllers/index.js (hunk 5 above)
Comment (abhilashlr) [reply app/controllers/index.js:20]: Oh yeah, that makes sense @ijlee2! But, I see there could be a catch where if a user's name is empty, what should we do on that data. Probably I'd like to take this in a separate PR.

File: app/controllers/index.js (hunk 6 above)
Comment (abhilashlr) [reply app/controllers/index.js:37]: On the other hand, I feel, we should keep it in sync with what the GitHub API sends in response? That way, it is easier to make changes in the future. I'd rather move this to ember-data models probably.

File: app/controllers/index.js (hunk 2 above)
Comment (abhilashlr) [reply app/controllers/index.js]: Yeah, we can!

File: app/controllers/index.js (hunk 5 above)
Comment (ijlee2) [reply app/controllers/index.js:20]: Yeah, I can see the preference to keep the names as given.

I was thinking we could remove Ember Data some time, heh. We only use 1 model, `github-pull`, for reading data. I think we could use `fetch` instead and simplify the app.

Comment (abhilashlr) [reply app/controllers/index.js:20]: Yep, agreed
